src/databricks/labs/ucx/assessment/jobs.py

```diff
@@ -104,3 +116,221 @@ def snapshot(self) -> Iterable[JobInfo]:
+        # Any fs-related configuration, e.g. sts or fs.azure.account.key
+        _failures = []
+        if task_cluster.aws_attributes is not None and task_cluster.aws_attributes.instance_profile_arn is not None:
+            _failures.append("AWS instance profile in the configuration")
+        if task_cluster.spark_conf is not None:
+            azure_failures = self.check_spark_conf(task_cluster.spark_conf, "Task cluster")
+            _failures.extend(azure_failures)
+            for key_regexp in _FS_LEVEL_KEY_REGEXP:
+                for key in task_cluster.spark_conf.keys():
+                    if re.match(key_regexp, key):
+                        _failures.append(f"fs-related configuration: {key}")
+        return _failures
+
+    def _assess_job_runs(self, submit_runs: list[BaseRun]) -> Iterable[SubmitRunInfo]:
+        """
+        Assessment logic:
+        1. For each submit run, we analyze all tasks inside this run.
+        2. Per each task, we calculate a unique hash based on the _retrieve_hash_values_from_task function
+        3. Then we coalesce all task hashes into a single hash for the submit run
+        4. Coalesce all runs under the same hash into a single pseudo-job
+        5. Return a list of pseudo-jobs with their assessment results
+        """
+        result: dict[str, SubmitRunInfo] = {}
+        runs_per_hash: dict[str, list[int | None]] = {}
+        for submit_run in submit_runs:
+            hashable_items = []
+            all_tasks: list[RunTask] = submit_run.tasks if submit_run.tasks is not None else []
+            failures_per_task: dict[str, list[str]] = {}
+
+            for task in sorted(all_tasks, key=lambda x: x.task_key if x.task_key is not None else ""):
+                # we recognize a task as a potentially incompatible one if:
+                # 1. cluster is not configured with data security mode
+                # 2. cluster's spark version is greater than 11.3 or is a custom image
+
+                hashable_items.extend(self._retrieve_hash_values_from_task(task))
+
+                task_cluster = _get_job_cluster_from_task(task)
+                if task_cluster is None:
+                    continue
+                if task_cluster.data_security_mode is None:
+                    dbr_info = parse_spark_version(task_cluster.spark_version)
+                    if not dbr_info.dbr_version:
+                        # we can skip the task since for
+                        # clusters without a numeric version data_security_mode will not be opted-in
+                        continue
+                    if parse_spark_version(task_cluster.spark_version).requires_submit_runs_check:
+                        # this task has version >= 11.3, we need to check it
+                        _task_key = task.task_key if task.task_key is not None else ""
+                        failures_per_task[_task_key] = self._check_task_for_failures(task_cluster)
```

Comment:
Method is too long and nesting is too deep, hard to review

Reply:
agree, simplified